File: LeetCode/Python/Easy/169-Majority-Element/solution.py
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        result: int=0

        n: int=len(nums)

        majority: int=(n // 2) + 1

        #Assume that the sign of the majority element is positive
        sign: int=1

        #Iterate through all 32 possible bit positions in an integer
        for bit_position in range(0, 32):
            
            #Keep track of the total number of 1 bits in each position
            set_bits: int=0

            #For each number
            for number in nums:
                
                #Get the bit in position bit_position
                #We take the absolute value of each number to make each number positive
                bit=(abs(number) >> bit_position) & 1

                #If the bit in position bit_position is a 1
                if bit==1:
                    set_bits+=1

            #If the total number of set_bits in position bit_position is at least majority
            if set_bits >= majority:
                
                #If the number is negative, then the majority element will also be negative
                if number < 0:
                    sign=-1
                
                #Set the bit in position bit_position to 1 since the majority element will have a 1 in the bit position bit_position
                result=result | (1 << bit_position)

        #Multiply the majority element by the sign in-case it is negative
        result*=sign

        return result

class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        result: int=0

        n: int=len(nums)

        majority: int=(n // 2) + 1

        hashmap: dict[int, int]=dict()

        for number in nums:
            if number not in hashmap:
                hashmap[number]=0
            
            hashmap[number]+=1

        for number, frequency in hashmap.items():
            if frequency >= majority:
                result=number

                break

        return result

File: LeetCode/Rust/Easy/169-Majority-Element/solution.rs
use std::collections::{HashMap};

impl Solution
{
    pub fn majority_element(nums: Vec<i32>) -> i32
    {
        let mut result: i32=0;

        let n: usize=nums.len();

        let mut hashmap: HashMap<i32, usize>=HashMap::new();

        for number in nums.iter()
        {
            hashmap.entry(*number).and_modify(|frequency| *frequency+=1).or_insert(1);
        }

        let target_frequency: usize=n / 2;

        for (number, frequency) in hashmap.iter()
        {
            if *frequency > target_frequency
            {
                result=*number;

                break;
            }
        }

        result
    }
}

impl Solution
{
    pub fn majority_element(nums: Vec<i32>) -> i32
    {
        let mut result: i32=0;

        let n: usize=nums.len();

        let mut hashmap: HashMap<i32, usize>=HashMap::new();

        nums.iter().for_each(|&number| {hashmap.entry(number).and_modify(|frequency| *frequency+=1).or_insert(1);});

        let target_frequency: usize=n / 2;

        for (number, frequency) in hashmap.iter()
        {
            if *frequency > target_frequency
            {
                result=*number;

                break;
            }
        }

        result
    }
}

impl Solution
{
    pub fn majority_element(nums: Vec<i32>) -> i32
    {
        let mut result: i32=0;

        let n: usize=nums.len();

        let majority: usize=(n / 2) + 1;

        let mut set_bits_per_position: Vec<usize>=vec![0; 32];

        let sign: i32=1;

        for bit_position in 0..32
        {
            let mut set_bits: usize=0;

            for number in nums.iter()
            {
                let bit: usize=((number >> bit_position) & 1) as usize;

                if bit==1
                {
                    set_bits+=1;
                }
            }

            set_bits_per_position[bit_position]+=set_bits;
        }

        for bit_position in 0..32
        {
            if set_bits_per_position[bit_position] >= majority
            {
                result=result | (1 << bit_position);
            }
        }

        result
    }
}